### ISSUE TITLE
Reference file spec update issues

### ISSUE THREAD

**Issue by jaytmiller** [spacetelescope/jwst#2379]
I propose opening a blanket issue for Q&A for the updated reference file spec.

**Comment by jaytmiller** [comment]
Generic Keywords

During the spec update the section on boiler plate keywords either disappeared or became harder to find.  Keywords like AUTHOR, USEAFTER, DATE, etc.   Some sections (AMI, TSO_PHOT) do define them but there are advantages to defining them generically as well if that section was indeed dropped.

**Comment by jaytmiller** [comment]
ASSIGN_WCS  - DISTORTION
NIRCAM is matching on both EXP_TYPE, DETECTOR, CHANNEL, PUPIL and FILTER.   PUPIL looks dominant relative to FILTER.   FILTER appears to be unused.
NIRISS is matching only on EXP_TYPE.

**Comment by jaytmiller** [comment]
ASSIGN_WCS WAVELENGTHRANGE
NIRCAM is matching on EXP_TYPE, PUPIL, MODULE.

**Comment by jaytmiller** [comment]
DQ ARRAY BIT DEFINITIONS
Thus far CRDS has done no checking of DQ_DEF bit definitions other than the extension is present and has the required table and columns.   However,  I noticed in the spec for background subtraction that data DQ and reference DQ are combined with a bitwise or operation.   The implications of the bitwise or is that DQ_DEF values either need to be universally consistent (and hence,  why define them file-by-file?) or more complex combination code is needed to normalize the data and reference DQ arrays to the same bit definitions prior to or'ing.

**Comment by hbushouse** [comment]
Bit definitions contained in the DQ_DEF table extension are automatically remapped onto the standard DQ bit definitions when reference files are loaded into a data model, by calls to the ``dynamic_mask`` function that are included in each reference file data model. Hence they can have unique mappings in each ref file.

**Comment by jaytmiller** [comment]
BARSHADOW
Intro typos.

**Comment by jaytmiller** [comment]
CUBE BUILDING CUBEPAR
retype is named cubepar,  not cubepars, correct spec.

To define checks in CRDS,  spec should literally name extensions, columns, and column type.   (Literal means e.g. don't say GRATING when the file uses DISPERSER.)

MIRI has 3 table extensions with 6 columns. column names are not identical.
NIRSPEC has 2 table extensions with 6 columns with differing names,  3 table extensions with 5 columns with identical names.

FWIW:  Both MIRI and NIRSPEC may be using f4 and f8 column formats inconsistently,  precision may vary by extension for same column name.

Need to decide yay-or-nay on CRDS checks for: extension/table name, column names,  column formats, lookup/mode columns.

**Comment by jaytmiller** [comment]
CUBE BUILDING RESOL
Need to decide yay-or-nay on CRDS checks for: extension/table name, column names, column formats, lookup/mode columns.
[resol_info.txt](https://github.com/spacetelescope/jwst/files/2294870/resol_info.txt)

**Comment by jaytmiller** [comment]
FLAT
Matching criteria: NOTE: all MIRI EXP_TYPE need to define BAND,  even if only N/A.
NIRISS is also matching on PUPIL.

**Comment by jaytmiller** [comment]
GAIN
.tpn's exist for all instruments checking a SCI array.  The spec says only the GAINFACT keyword matters.  Propose changing .tpn's to check only GAINFACT / META.EXPOSURE.GAIN_FACTOR limited to range 1.0:10.0 as discussed for NIRSPEC.

**Comment by jaytmiller** [comment]
IPC 
Spec says center pixel values > 1.   Some references (FGS, others?) using at least some degenerate kernels with center 1.0 and zeroes.

**Comment by jaytmiller** [comment]
LINEARITY
No documentation on NIRSPEC IRS2 (256,2048) striping.  No docs on array dimensions in general, particularly with respect to reference pixels and YMAX=3200.

**Comment by jaytmiller** [comment]
PATHLOSS
Needs second round of editing,  paritcularly FIXEDSLIT definition.

**Comment by jaytmiller** [comment]
TRAPPARS

Trappars section should mention EXTNAME=TRAPPARS

**Comment by jaytmiller** [comment]
REFPIX
Some existing references define IRS2 and IRS2ALPHA extensions,  while others define IRS2 and DQ extensions.   Neither IRS2ALPHA nor DQ is mentioned.  There's no refpix schema.   (Not sure how to interpret "complex coefficients",  whether that's A+Bi or both A and B are supposed to be complex.  (They're all float32.))

**Comment by jaytmiller** [comment]
RESET
Defined as N/A for all types.  Remove?

**Comment by jaytmiller** [comment]
RSCD
Add comment about MIRI-only?

**Comment by jaytmiller** [comment]
Stray Light Mask

Spec should mention reftype name straymask.

**Comment by jaytmiller** [comment]
SUPERBIAS
Spec should comment on no MIRI?
Spec should mention NIRSPEC matching includes SUBSTRT1/2 and SUBSIZE1/2.

**Comment by jaytmiller** [comment]
TSOPHOT
Spec includes boilerplate keyword defs which should be factored out.

**Comment by jaytmiller** [comment]
EXTRACT2D
Typo NRS_FIXEDSLT vs. NRS_FIXEDSLIT.

**Comment by stscicrawford** [comment]
cc: @rizeladiaz

**Comment by stscicrawford** [comment]
Generic keywords have been added in #2385

**Comment by jaytmiller** [comment]
I marked sub-issues I believe are resolved by my pull request as X.
I marked sub-issues which still require input as XXX.

Overall,  I made Rossy's generic reference file info available as packages references_general and referenced it everywhere from the include file includes/standard_keywords.rst.   

I also replaced inline explanations of DQ_DEF with includes/dq_def.rst which further references Rossy's exhaustive list of DQ bit definitions (in a perfect world).   (I don't think I saw enough detail anywhere to make inline DQ_DEF explanations relevant.   The entire table seems like a necessary evil,  enabling non-standard definitions of common concepts.   I suspect some references have non-standard definitions which Howard indicated are normalized somehow.)

I also added localized selection criteria include files in many places to generate tables which spell out which instruments are currently supported and if/how their selection criteria differ.

For cube_build,  I added experimental array format tables.
